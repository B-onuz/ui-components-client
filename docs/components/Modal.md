# Modal (Component)

Componente de Modal. Utilizado em conjunto com o hook `useModal`.

[Documentação do hook useModal](/docs/hooks/useModal.md)

## Como utilizar

```js
import Modal from 'ui-components/lib/components/Modal'
```

```jsx
import React, { useEffect, useState } from 'react'
import Modal from 'ui-components/lib/components/Modal'
import useModal from 'ui-components/lib/hooks/useModal'

const ModalExample = () => {
  const { isOpen, openModal, closeModal, params: modalParams } = useModal()

  return (
    <>
      <Button onClick={() => openModal({ id })} m={10} color={'primary'}>
        Abrir modal
      </Button>
      <Modal open={isOpen} onClose={closeModal} header={<h3>Hello id: {(modalParams || {}).id}</h3>} footer={<small>Footer</small>}>
        <h1>lorem ipsum modal 3</h1>
        <p>Lorem ipsum dolor sit amet</p>
      </Modal>
    </>
  )
}
export default ModalExample
```

## props

### `boxProps`

Utilizado para passar propriedades para o componente Box da modal.

**ex:**

```jsx
<Modal boxProps={{ px: 20, py: 10 }}>
  <p> Velit non duis ut elit mollit elit eu consequat velit.</p>
</Modal>
```

### `contentModalProps`

Utilizado para passar propriedades para o componente que envolve o box da modal.

**ex:**

```jsx
<Modal contentModalProps={{ width: 1024 }}>
  <p> Velit non duis ut elit mollit elit eu consequat velit.</p>
</Modal>
```

### `header`

Slot para componente ou texto. Utilizado para renderizar um cabeçalho.

**ex:**

```jsx
<Modal header={<h1>Título</h1>}>
  <p> Velit non duis ut elit mollit elit eu consequat velit.</p>
</Modal>
```

### `footer`

Slot para componente ou texto. Utilizado para renderizar um rodapé.

**ex:**

```jsx
<Modal footer={<footer>rodapé</footer>}>
  <p> Velit non duis ut elit mollit elit eu consequat velit.</p>
</Modal>
```
