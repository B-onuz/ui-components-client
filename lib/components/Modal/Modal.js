"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Box = _interopRequireDefault(require("../Box"));

var _reactSpring = require("react-spring");

var _styledSystem = require("styled-system");

var _this = void 0,
    _jsxFileName = "/Users/master/projects/bonuz/ui-components/src/components/Modal/Modal.js";

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _templateObject7() {
  var data = _taggedTemplateLiteralLoose(["\n  position: absolute;\n  z-index: -1;\n  background: rgba(0, 0, 0, 0.3);\n  width: 100%;\n  height: 100%;\n  top: 0;\n  right: 0;\n  bottom: 0;\n  left: 0;\n  cursor: pointer;\n  border: none;\n  outline: none;\n"]);

  _templateObject7 = function _templateObject7() {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = _taggedTemplateLiteralLoose(["\n  position: fixed;\n  top: 0;\n  right: 0;\n  bottom: 0;\n  left: 0;\n  width: 100%;\n  height: 100%;\n  z-index: 9999999;\n  display: flex;\n  align-items: center;\n  justify-content: center;\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = _taggedTemplateLiteralLoose([""]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteralLoose([""]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteralLoose(["\n  overflow: auto;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose(["\n  box-sizing: border-box;\n  max-height: 90vh;\n  display: flex;\n  flex-direction: column;\n  padding: ", ";\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n  width: auto;\n  max-width: 90%;\n  margin: auto;\n  ", "\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteralLoose(strings, raw) { if (!raw) { raw = strings.slice(0); } strings.raw = raw; return strings; }

var ContentModal = (0, _styledComponents["default"])(_reactSpring.animated.div)(_templateObject(), (0, _styledSystem.compose)(_styledSystem.space, _styledSystem.layout));
var BoxModal = (0, _styledComponents["default"])(_Box["default"])(_templateObject2(), function (_ref) {
  var spacing = _ref.spacing;
  return spacing ? spacing : '20px';
});

var BoxModalContent = _styledComponents["default"].div(_templateObject3());

var BoxModalHeader = _styledComponents["default"].div(_templateObject4());

var BoxModalFooter = _styledComponents["default"].div(_templateObject5());

var Wrapper = _styledComponents["default"].div(_templateObject6());

var Shadow = _styledComponents["default"].button(_templateObject7());

var useRootModal = function useRootModal() {
  var _useState = (0, _react.useState)(),
      $rootModal = _useState[0],
      setRootModal = _useState[1];

  (0, _react.useEffect)(function () {
    var rootModal = document.getElementById('modal-root');

    if (rootModal) {
      setRootModal(rootModal);
    } else {
      var rootEl = document.createElement('div');
      rootEl.id = 'modal-root';
      setRootModal(rootEl);
      document.body.appendChild(rootEl);
    }
  });
  return {
    $rootModal: $rootModal
  };
};

var Modal = function Modal(_ref2) {
  var children = _ref2.children,
      open = _ref2.open,
      onClose = _ref2.onClose,
      boxProps = _ref2.boxProps,
      header = _ref2.header,
      footer = _ref2.footer,
      contentModalProps = _ref2.contentModalProps,
      spacing = _ref2.spacing,
      rest = _objectWithoutPropertiesLoose(_ref2, ["children", "open", "onClose", "boxProps", "header", "footer", "contentModalProps", "spacing"]);

  var _useRootModal = useRootModal(),
      $rootModal = _useRootModal.$rootModal;

  var transitions = (0, _reactSpring.useTransition)(open, null, {
    from: {
      transform: 'translate3d(0,-80px,0)',
      opacity: 0
    },
    enter: {
      transform: 'translate3d(0,0px,0)',
      opacity: 1
    },
    leave: {
      transform: 'translate3d(0,-80px,0)',
      opacity: 0
    }
  });
  var opacityTransitions = (0, _reactSpring.useTransition)(open, null, {
    from: {
      opacity: 0
    },
    enter: {
      opacity: 1
    },
    leave: {
      opacity: 0
    }
  });
  if (!$rootModal) return null;
  return _reactDom["default"].createPortal(opacityTransitions.map(function (_ref3) {
    var item = _ref3.item,
        key = _ref3.key;
    return item && /*#__PURE__*/_react["default"].createElement(Wrapper, _extends({}, rest, {
      key: key,
      role: "dialog",
      "aria-modal": "true",
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 97,
        columnNumber: 11
      }
    }), transitions.map(function (_ref4) {
      var item = _ref4.item,
          key = _ref4.key,
          props = _ref4.props;
      return item && /*#__PURE__*/_react["default"].createElement(ContentModal, _extends({
        key: key,
        style: props
      }, contentModalProps, {
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 101,
          columnNumber: 19
        }
      }), /*#__PURE__*/_react["default"].createElement(BoxModal, _extends({}, boxProps, {
        spacing: spacing,
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 102,
          columnNumber: 21
        }
      }), /*#__PURE__*/_react["default"].createElement(BoxModalHeader, {
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 103,
          columnNumber: 23
        }
      }, header), /*#__PURE__*/_react["default"].createElement(BoxModalContent, {
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 104,
          columnNumber: 23
        }
      }, children), /*#__PURE__*/_react["default"].createElement(BoxModalFooter, {
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 105,
          columnNumber: 23
        }
      }, footer)));
    }), opacityTransitions.map(function (_ref5) {
      var item = _ref5.item,
          key = _ref5.key,
          props = _ref5.props;
      return item && /*#__PURE__*/_react["default"].createElement(Shadow, {
        key: key,
        style: props,
        onClick: onClose,
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 111,
          columnNumber: 71
        }
      });
    }));
  }), $rootModal);
};

var _default = Modal;
exports["default"] = _default;
module.exports = exports.default;