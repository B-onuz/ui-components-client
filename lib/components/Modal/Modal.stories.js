'use strict'

function _typeof(obj) {
  '@babel/helpers - typeof'
  if (typeof Symbol === 'function' && typeof Symbol.iterator === 'symbol') {
    _typeof = function _typeof(obj) {
      return typeof obj
    }
  } else {
    _typeof = function _typeof(obj) {
      return obj && typeof Symbol === 'function' && obj.constructor === Symbol && obj !== Symbol.prototype ? 'symbol' : typeof obj
    }
  }
  return _typeof(obj)
}

exports.__esModule = true
exports.useModalHook = exports.MultipleModals = exports['default'] = void 0

var _react = _interopRequireWildcard(require('react'))

var _Modal = _interopRequireDefault(require('./Modal'))

var _Button = _interopRequireDefault(require('../Button'))

var _useModal2 = _interopRequireDefault(require('../../hooks/useModal'))

var _this = void 0,
  _jsxFileName = '/home/junior/Fontes/ui-components-client/src/components/Modal/Modal.stories.js'

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj }
}

function _getRequireWildcardCache() {
  if (typeof WeakMap !== 'function') return null
  var cache = new WeakMap()
  _getRequireWildcardCache = function _getRequireWildcardCache() {
    return cache
  }
  return cache
}

function _interopRequireWildcard(obj) {
  if (obj && obj.__esModule) {
    return obj
  }
  if (obj === null || (_typeof(obj) !== 'object' && typeof obj !== 'function')) {
    return { default: obj }
  }
  var cache = _getRequireWildcardCache()
  if (cache && cache.has(obj)) {
    return cache.get(obj)
  }
  var newObj = {}
  var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor
  for (var key in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, key)) {
      var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null
      if (desc && (desc.get || desc.set)) {
        Object.defineProperty(newObj, key, desc)
      } else {
        newObj[key] = obj[key]
      }
    }
  }
  newObj['default'] = obj
  if (cache) {
    cache.set(obj, newObj)
  }
  return newObj
}

var _default = {
  title: 'Modal',
  component: _Modal['default'],
}
exports['default'] = _default

var MultipleModals = function MultipleModals() {
  var _useState = (0, _react.useState)([false, false, false]),
    modalOpen = _useState[0],
    setModalOpen = _useState[1]

  var openModal = function openModal(key) {
    setModalOpen(
      modalOpen.map(function (value, index) {
        if (index === key) return true
        return value
      })
    )
  }

  var closeModal = function closeModal(key) {
    setModalOpen(
      modalOpen.map(function (value, index) {
        if (index === key) return false
        return value
      })
    )
  }

  return /*#__PURE__*/ _react['default'].createElement(
    _react['default'].Fragment,
    null,
    /*#__PURE__*/ _react['default'].createElement(
      _Button['default'],
      {
        onClick: function onClick() {
          return openModal(0)
        },
        m: 10,
        color: 'primary',
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 34,
          columnNumber: 7,
        },
      },
      'Abrir modal 1'
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _Button['default'],
      {
        onClick: function onClick() {
          return openModal(1)
        },
        m: 10,
        color: 'primary',
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 37,
          columnNumber: 7,
        },
      },
      'Abrir modal 2'
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _Button['default'],
      {
        onClick: function onClick() {
          return openModal(2)
        },
        m: 10,
        color: 'primary',
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 40,
          columnNumber: 7,
        },
      },
      'Abrir modal 3'
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _Modal['default'],
      {
        open: modalOpen[0],
        onClose: function onClose() {
          return closeModal(0)
        },
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 43,
          columnNumber: 7,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(
        'h1',
        {
          __self: _this,
          __source: {
            fileName: _jsxFileName,
            lineNumber: 44,
            columnNumber: 9,
          },
        },
        'lorem ipsum modal 1'
      ),
      /*#__PURE__*/ _react['default'].createElement(
        _Button['default'],
        {
          onClick: function onClick() {
            return openModal(1)
          },
          m: 10,
          color: 'primary',
          __self: _this,
          __source: {
            fileName: _jsxFileName,
            lineNumber: 45,
            columnNumber: 9,
          },
        },
        'Abrir modal 2'
      ),
      /*#__PURE__*/ _react['default'].createElement(
        _Button['default'],
        {
          onClick: function onClick() {
            return openModal(2)
          },
          m: 10,
          color: 'primary',
          __self: _this,
          __source: {
            fileName: _jsxFileName,
            lineNumber: 48,
            columnNumber: 9,
          },
        },
        'Abrir modal 3'
      )
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _Modal['default'],
      {
        open: modalOpen[1],
        onClose: function onClose() {
          return closeModal(1)
        },
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 52,
          columnNumber: 7,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(
        'h1',
        {
          __self: _this,
          __source: {
            fileName: _jsxFileName,
            lineNumber: 53,
            columnNumber: 9,
          },
        },
        'lorem ipsum modal 2'
      ),
      /*#__PURE__*/ _react['default'].createElement(
        _Button['default'],
        {
          onClick: function onClick() {
            return openModal(0)
          },
          m: 10,
          color: 'primary',
          __self: _this,
          __source: {
            fileName: _jsxFileName,
            lineNumber: 54,
            columnNumber: 9,
          },
        },
        'Abrir modal 1'
      ),
      /*#__PURE__*/ _react['default'].createElement(
        _Button['default'],
        {
          onClick: function onClick() {
            return openModal(2)
          },
          m: 10,
          color: 'primary',
          __self: _this,
          __source: {
            fileName: _jsxFileName,
            lineNumber: 57,
            columnNumber: 9,
          },
        },
        'Abrir modal 3'
      )
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _Modal['default'],
      {
        open: modalOpen[2],
        spacing: '40px 60px',
        onClose: function onClose() {
          return closeModal(2)
        },
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 61,
          columnNumber: 7,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(
        'h1',
        {
          __self: _this,
          __source: {
            fileName: _jsxFileName,
            lineNumber: 62,
            columnNumber: 9,
          },
        },
        'lorem ipsum modal 3'
      ),
      /*#__PURE__*/ _react['default'].createElement(
        _Button['default'],
        {
          onClick: function onClick() {
            return openModal(0)
          },
          m: 10,
          color: 'primary',
          __self: _this,
          __source: {
            fileName: _jsxFileName,
            lineNumber: 63,
            columnNumber: 9,
          },
        },
        'Abrir modal 1'
      ),
      /*#__PURE__*/ _react['default'].createElement(
        _Button['default'],
        {
          onClick: function onClick() {
            return openModal(1)
          },
          m: 10,
          color: 'primary',
          __self: _this,
          __source: {
            fileName: _jsxFileName,
            lineNumber: 66,
            columnNumber: 9,
          },
        },
        'Abrir modal 2'
      )
    )
  )
}

exports.MultipleModals = MultipleModals

var useModalHook = function useModalHook() {
  var _useModal = (0, _useModal2['default'])(),
    isOpen = _useModal.isOpen,
    openModal = _useModal.openModal,
    closeModal = _useModal.closeModal,
    modalParams = _useModal.params

  return /*#__PURE__*/ _react['default'].createElement(
    _react['default'].Fragment,
    null,
    /*#__PURE__*/ _react['default'].createElement(
      _Button['default'],
      {
        onClick: function onClick() {
          return openModal({
            id: '#1',
          })
        },
        m: 10,
        color: 'primary',
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 79,
          columnNumber: 7,
        },
      },
      'Abrir modal'
    ),
    /*#__PURE__*/ _react['default'].createElement(
      _Modal['default'],
      {
        contentModalProps: {
          width: 1024,
        },
        boxProps: {
          p: 40,
        },
        open: isOpen,
        onClose: closeModal,
        header: /*#__PURE__*/ _react['default'].createElement(
          'h3',
          {
            __self: _this,
            __source: {
              fileName: _jsxFileName,
              lineNumber: 87,
              columnNumber: 17,
            },
          },
          'Hello id: ',
          (modalParams || {}).id
        ),
        footer: /*#__PURE__*/ _react['default'].createElement(
          'small',
          {
            __self: _this,
            __source: {
              fileName: _jsxFileName,
              lineNumber: 88,
              columnNumber: 17,
            },
          },
          'Footer'
        ),
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 82,
          columnNumber: 7,
        },
      },
      /*#__PURE__*/ _react['default'].createElement(
        'p',
        {
          style: {
            maxWidth: 320,
          },
          __self: _this,
          __source: {
            fileName: _jsxFileName,
            lineNumber: 90,
            columnNumber: 9,
          },
        },
        'Magna ullamco incididunt laborum laboris. Qui Lorem nisi laboris minim elit ipsum magna aute commodo quis. Et minim nostrud dolore reprehenderit aliqua non ad quis et. Commodo excepteur proident Lorem laboris non qui pariatur duis id amet. Mollit minim commodo magna mollit dolor irure dolor sint Lorem nostrud reprehenderit ea. Elit laborum in eiusmod consectetur consequat dolore. Sit nostrud deserunt officia laborum Lorem sint officia. Adipisicing veniam aliquip ex reprehenderit tempor velit tempor amet deserunt. Ad veniam minim exercitation officia. Excepteur laborum tempor qui non elit id quis esse enim. Anim do sint fugiat cupidatat dolor officia incididunt incididunt laboris deserunt irure. Velit in dolor est ex aute tempor sunt laborum cupidatat velit amet. Deserunt sunt eu voluptate nisi dolore aliquip cillum ea magna culpa cillum Lorem. Esse occaecat mollit consectetur non velit laborum qui anim occaecat fugiat amet adipisicing. Et ut voluptate duis consectetur dolor eu cupidatat. Voluptate in ipsum do est dolore magna est velit magna in irure. Excepteur proident sit ut Lorem et aute ut. Adipisicing id cillum elit reprehenderit nisi deserunt consequat. Elit duis ad commodo incididunt est. Voluptate sint Lorem minim voluptate in.'
      )
    )
  )
}

exports.useModalHook = useModalHook
