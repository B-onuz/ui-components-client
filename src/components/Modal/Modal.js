import React, { useEffect, useState } from 'react'
import ReactDOM from 'react-dom'
import styled from 'styled-components'
import Box from '../Box'
import { animated, useTransition } from 'react-spring'
import { space, layout, compose } from 'styled-system'

const ContentModal = styled(animated.div)`
  width: auto;
  max-width: 90%;
  margin: auto;
  ${compose(space, layout)}
`

const BoxModal = styled(Box)`
  box-sizing: border-box;
  max-height: 90vh;
  display: flex;
  flex-direction: column;
  padding: ${({ spacing }) => (spacing ? spacing : '20px')};
`

const BoxModalContent = styled.div`
  overflow: auto;
`

const BoxModalHeader = styled.div``
const BoxModalFooter = styled.div``

const Wrapper = styled.div`
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 9999999;
  display: flex;
  align-items: center;
  justify-content: center;
`

const Shadow = styled.button`
  position: absolute;
  z-index: -1;
  background: rgba(0, 0, 0, 0.3);
  width: 100%;
  height: 100%;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  cursor: pointer;
  border: none;
  outline: none;
`

const useRootModal = () => {
  const [$rootModal, setRootModal] = useState()

  useEffect(() => {
    const rootModal = document.getElementById('modal-root')
    if (rootModal) {
      setRootModal(rootModal)
    } else {
      const rootEl = document.createElement('div')
      rootEl.id = 'modal-root'
      setRootModal(rootEl)
      document.body.appendChild(rootEl)
    }
  })

  return { $rootModal }
}

const Modal = ({ children, open, onClose, boxProps, header, footer, contentModalProps, spacing, ...rest }) => {
  const { $rootModal } = useRootModal()
  const transitions = useTransition(open, null, {
    from: { transform: 'translate3d(0,-80px,0)', opacity: 0 },
    enter: { transform: 'translate3d(0,0px,0)', opacity: 1 },
    leave: { transform: 'translate3d(0,-80px,0)', opacity: 0 },
  })

  const opacityTransitions = useTransition(open, null, {
    from: { opacity: 0 },
    enter: { opacity: 1 },
    leave: { opacity: 0 },
  })

  if (!$rootModal) return null

  return ReactDOM.createPortal(
    opacityTransitions.map(
      ({ item, key }) =>
        item && (
          <Wrapper {...rest} key={key} role="dialog" aria-modal="true">
            {transitions.map(
              ({ item, key, props }) =>
                item && (
                  <ContentModal key={key} style={props} {...contentModalProps}>
                    <BoxModal {...boxProps} spacing={spacing}>
                      <BoxModalHeader>{header}</BoxModalHeader>
                      <BoxModalContent>{children}</BoxModalContent>
                      <BoxModalFooter>{footer}</BoxModalFooter>
                    </BoxModal>
                  </ContentModal>
                )
            )}

            {opacityTransitions.map(({ item, key, props }) => item && <Shadow key={key} style={props} onClick={onClose} />)}
          </Wrapper>
        )
    ),
    $rootModal
  )
}

export default Modal
