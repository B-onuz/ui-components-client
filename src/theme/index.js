const colors = {
  red: '#db2828',
  lightRed: '#ef4d5a',
  mediumRed: '#FF4F5D',
  darkRed: '#9c2543',
  fullDarkRed: '#550e33',
  orange: '#f2711c',
  yellow: '#fbbd08',
  olive: '#b5cc18',
  green: '#21BA45',
  teal: '#00b5ad',
  blue: '#2185d0',
  facebookBlue: '#3B5998',
  blueLink: '#4285F4',
  violet: '#6435c9',
  purple: '#a333c8',
  pink: '#e03997',
  brown: '#a5673f',
  grey: '#767676',
  textGrey: '#8D8D8D',
  lightGrey: '#6D6D6D',
  mediumGrey: '#838383',
  darkGrey: '#9A9A9A',
  boldGray: '#55555A',
  black: '#1b1c1d',
  white: '#fdfdfd',
  warning: '#7A4D05',
  success: '#1E561F',
  info: '#0E566C',
  danger: '#973937',
  cyan: '#1CB5AD',
  disabledGrey: '#ffffff1f',
  disabledGreyText: '#ffffff4d',
}

colors.primary = colors.lightRed
colors.secondary = colors.fullDarkRed
colors.default = '#e0e1e2'

const breakpoints = ['40em', '52em', '64em', '80em']

breakpoints.sm = breakpoints[0]
breakpoints.md = breakpoints[1]
breakpoints.lg = breakpoints[2]
breakpoints.xl = breakpoints[3]

const fontSizes = [12, 14, 16, 20, 24, 32]

export default {
  fontFamily: `'Lato', sans-serif`,
  colors,
  fontSizes,
  breakpoints,
}
