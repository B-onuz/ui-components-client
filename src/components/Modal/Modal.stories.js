import React, { useState } from 'react'
import Modal from './Modal'
import Button from '../Button'
import useModal from '../../hooks/useModal'

export default {
  title: 'Modal',
  component: Modal,
}

export const MultipleModals = () => {
  const [modalOpen, setModalOpen] = useState([false, false, false])

  const openModal = (key) => {
    setModalOpen(
      modalOpen.map((value, index) => {
        if (index === key) return true
        return value
      })
    )
  }

  const closeModal = (key) => {
    setModalOpen(
      modalOpen.map((value, index) => {
        if (index === key) return false
        return value
      })
    )
  }

  return (
    <>
      <Button onClick={() => openModal(0)} m={10} color={'primary'}>
        Abrir modal 1
      </Button>
      <Button onClick={() => openModal(1)} m={10} color={'primary'}>
        Abrir modal 2
      </Button>
      <Button onClick={() => openModal(2)} m={10} color={'primary'}>
        Abrir modal 3
      </Button>
      <Modal open={modalOpen[0]} onClose={() => closeModal(0)}>
        <h1>lorem ipsum modal 1</h1>
        <Button onClick={() => openModal(1)} m={10} color={'primary'}>
          Abrir modal 2
        </Button>
        <Button onClick={() => openModal(2)} m={10} color={'primary'}>
          Abrir modal 3
        </Button>
      </Modal>
      <Modal open={modalOpen[1]} onClose={() => closeModal(1)}>
        <h1>lorem ipsum modal 2</h1>
        <Button onClick={() => openModal(0)} m={10} color={'primary'}>
          Abrir modal 1
        </Button>
        <Button onClick={() => openModal(2)} m={10} color={'primary'}>
          Abrir modal 3
        </Button>
      </Modal>
      <Modal open={modalOpen[2]} spacing="40px 60px" onClose={() => closeModal(2)}>
        <h1>lorem ipsum modal 3</h1>
        <Button onClick={() => openModal(0)} m={10} color={'primary'}>
          Abrir modal 1
        </Button>
        <Button onClick={() => openModal(1)} m={10} color={'primary'}>
          Abrir modal 2
        </Button>
      </Modal>
    </>
  )
}

export const useModalHook = () => {
  const { isOpen, openModal, closeModal, params: modalParams } = useModal()

  return (
    <>
      <Button onClick={() => openModal({ id: '#1' })} m={10} color={'primary'}>
        Abrir modal
      </Button>
      <Modal
        contentModalProps={{ width: 1024 }}
        boxProps={{ p: 40 }}
        open={isOpen}
        onClose={closeModal}
        header={<h3>Hello id: {(modalParams || {}).id}</h3>}
        footer={<small>Footer</small>}
      >
        <p style={{ maxWidth: 320 }}>
          Magna ullamco incididunt laborum laboris. Qui Lorem nisi laboris minim elit ipsum magna aute commodo quis. Et minim nostrud dolore
          reprehenderit aliqua non ad quis et. Commodo excepteur proident Lorem laboris non qui pariatur duis id amet. Mollit minim commodo magna
          mollit dolor irure dolor sint Lorem nostrud reprehenderit ea. Elit laborum in eiusmod consectetur consequat dolore. Sit nostrud deserunt
          officia laborum Lorem sint officia. Adipisicing veniam aliquip ex reprehenderit tempor velit tempor amet deserunt. Ad veniam minim
          exercitation officia. Excepteur laborum tempor qui non elit id quis esse enim. Anim do sint fugiat cupidatat dolor officia incididunt
          incididunt laboris deserunt irure. Velit in dolor est ex aute tempor sunt laborum cupidatat velit amet. Deserunt sunt eu voluptate nisi
          dolore aliquip cillum ea magna culpa cillum Lorem. Esse occaecat mollit consectetur non velit laborum qui anim occaecat fugiat amet
          adipisicing. Et ut voluptate duis consectetur dolor eu cupidatat. Voluptate in ipsum do est dolore magna est velit magna in irure. Excepteur
          proident sit ut Lorem et aute ut. Adipisicing id cillum elit reprehenderit nisi deserunt consequat. Elit duis ad commodo incididunt est.
          Voluptate sint Lorem minim voluptate in.
        </p>
      </Modal>
    </>
  )
}
