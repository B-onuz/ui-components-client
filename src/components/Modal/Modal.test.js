import React from 'react'
import Modal from './Modal.js'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import theme from '../../theme'
import 'jest-styled-components'
import ReactDOM from 'react-dom'

describe('Component: Modal', () => {
  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((element, node) => {
      return element
    })
  })

  afterEach(() => {
    ReactDOM.createPortal.mockClear()
  })

  it('should render correctly when open is true', () => {
    const component = renderer.create(
      <ThemeProvider theme={theme}>
        <Modal open={true} onClose={() => {}}>
          <h1>lorem ipsum modal 3</h1>
          <p>Lorem ipsum dolor sit amet</p>
        </Modal>
      </ThemeProvider>
    )

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should render correctly when open is false', () => {
    const component = renderer.create(
      <ThemeProvider theme={theme}>
        <Modal open={false} onClose={() => {}}>
          <h1>lorem ipsum modal 3</h1>
          <p>Lorem ipsum dolor sit amet</p>
        </Modal>
      </ThemeProvider>
    )

    expect(component.toJSON()).toMatchSnapshot()
  })
})
