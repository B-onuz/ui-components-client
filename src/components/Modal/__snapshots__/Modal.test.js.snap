// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Modal should render correctly when open is false 1`] = `null`;

exports[`Component: Modal should render correctly when open is true 1`] = `null`;
